sw/device/tests/sim_dv/uart_tx_rx_test.c

```diff
@@ -0,0 +1,350 @@
+    0xc7, 0x33, 0xd1, 0x6a, 0xc7, 0xba, 0x78, 0x69};
+static size_t uart_rx_bytes_read;
+
+// Set our expectation & event indications of the interrupts we intend to
+// exercise in this test.
+static bool uart_irq_tx_watermark_fired;
+static bool uart_irq_rx_watermark_fired;
+static bool uart_irq_tx_empty_fired;
+static bool uart_irq_rx_overflow_fired;
+
+/**
+ * Send data out over TX without blocking.
+ *
+ * Non-blocking function that writes uart_tx_data bytes to be sent out of UART.
+ * It stops when the TX fifo is full. As and when the TX watermark interrupt
+ * fires, the test calls this function to resume sending the remaining bytes.
+ */
+void send_tx_data_nonblocking(void) {
+  size_t bytes_remaining = sizeof(uart_tx_data) - uart_tx_bytes_written;
+  // Return if we have no data to send.
+  if (bytes_remaining == 0)
+    return;
+  // Send the remaining uart_tx_data.
+  size_t bytes_written_isr = 0;
+  bool status =
+      dif_uart_bytes_send(&uart0, &uart_tx_data[uart_tx_bytes_written],
+                          bytes_remaining, &bytes_written_isr);
+  CHECK(status, "dif_uart_bytes_send failed");
+  uart_tx_bytes_written += bytes_written_isr;
+}
+
+/**
+ * Receive data in over RX without blocking.
+ *
+ * Non-blocking function that reads bytes received over UART RX port. It reads
+ * as many bytes as there are available in the RX FIFO and returns the number of
+ * bytes read, which is used the next time this function is called. Eventually,
+ * all UART_DATASET_SIZE bytes need to be read.
+ */
+void receive_rx_data_nonblocking(void) {
+  size_t bytes_remaining = sizeof(exp_uart_rx_data) - uart_rx_bytes_read;
+  // Return if we have no data to send.
+  if (bytes_remaining == 0)
+    return;
+  // Receive the remaining uart_rx_data.
+  size_t bytes_read_isr = 0;
+  bool status = dif_uart_bytes_receive(&uart0, bytes_remaining,
+                                       &uart_rx_data[uart_rx_bytes_read],
+                                       &bytes_read_isr);
+  CHECK(status, "dif_uart_bytes_receive failed");
+  uart_rx_bytes_read += bytes_read_isr;
+}
+
+/**
+ * Provides external irq handling for this test.
+ *
+ * This function overrides the default external irq handler in
+ * `sw/device/lib/handler.h`.
+ */
+void handler_irq_external(void) {
+  // Find which interrupt fired at PLIC by claiming it.
+  dif_plic_irq_id_t plic_irq_id;
+  bool status =
+      dif_plic_irq_claim(&plic0, kTopEarlgreyPlicTargetIbex0, &plic_irq_id);
+  CHECK(status, "dif_plic_irq_claim failed");
+
+  // Check if it is the right peripheral.
+  top_earlgrey_plic_peripheral_t peripheral = (top_earlgrey_plic_peripheral_t)
+      top_earlgrey_plic_interrupt_for_peripheral[plic_irq_id];
+  CHECK(peripheral == kTopEarlgreyPlicPeripheralUart,
+        "Only UART is expected to generate interrupts in this test!");
+
+  // Correlate the interrupt fired at PLIC with UART.
+  dif_uart_interrupt_t uart_irq = (dif_uart_interrupt_t)(
```

Comment:
This is a a pretty nice way of mapping the plic interrupts to uart, however I am slightly paranoid. Two reasons:
* I am not sure whether `top_earlgrey.sv` `intr_vector` signals are guaranteed to be in the same sequence as interrupts internally for the peripheral. It makes sense if they are, but then I think we have to explicitly state it somewhere in the documentation, and make it obvious for people to find (maybe OT UART doc?).
* `dif_uart_interrupt_t` are in the same sequence as they are defined in the specification - but I just used common sense, I don't think it was a hard requirement before now. But if to use this code, I think it has to be.

Alternative would be to have some sort of array struct mapping, similar to what is done for peripheral to irq id mapping in `"hw/top_earlgrey/sw/autogen/top_earlgrey.h"`.

Another alternative is to auto generate UART interrupt enums.

Would be nice to here some extra opinions on this.

Reply:
> I have changed the implementation in any case. It would be great to autogenerate the dif_uart_interrupt_t enum as well

I agree, I though I have opened an issue regarding it, but it seems I only mentioned it is one of the code reviews.